File: rust/src/model/domain/document.rs
use super::business_partner::BusinessPartner;
use super::{COGS_ACCOUNT, PAYABLE_ACCOUNT, Customer, Material, INVENTORY_ACCOUNT, INVENTORY_DIFF_ACCOUNT, User, Vendor, SALES_ACCOUNT, RECEIVABLE_ACCOUNT};
use chrono::NaiveDate;
use rust_decimal::Decimal;

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum DebtCredit {
    Debt,
    Credit,
}

impl DebtCredit {
    pub fn to_db_option(&self) -> &str {
        match self {
            DebtCredit::Debt => "D",
            DebtCredit::Credit => "C",
        }
    }
}

pub struct Document {
    pub posting_date: NaiveDate,
    pub doc_no: String,
    pub updated_by: User,
    pub lines: Vec<DocumentLine>,
}

impl Document {
    pub fn new_purchase_document(
        posting_date: NaiveDate,
        doc_no: String,
        material: Material,
        vendor: Vendor,
        price: Decimal,
        std_price: Decimal,
        quantity: Decimal,
        updated_by: User,
    ) -> Document {
        let amount = price * quantity;
        let std_amount = std_price * quantity;
        let diff_amount = amount - std_amount;

        let mut lines = vec![DocumentLine {
            account: INVENTORY_ACCOUNT,
            material: Some(material),
            business_partner: None,
            debt_credit: DebtCredit::Debt,
            amount: std_amount,
            quantity: Some(quantity),
        }];

        if diff_amount != Decimal::from(0) {
            let debt_credit = if diff_amount > Decimal::from(0) {
                DebtCredit::Debt
            } else {
                DebtCredit::Credit
            };
            let diff_line = DocumentLine {
                account: INVENTORY_DIFF_ACCOUNT,
                material: Some(material),
                business_partner: None,
                debt_credit,
                amount: diff_amount,
                quantity: None,
            };
            lines.push(diff_line);
        }

        lines.push(DocumentLine {
            account: PAYABLE_ACCOUNT,
            material: None,
            business_partner: Some(BusinessPartner::Vendor(vendor)),
            debt_credit: DebtCredit::Credit,
            amount: -amount,
            quantity: None,
        });

        Document {
            posting_date,
            doc_no,
            updated_by,
            lines,
        }
    }

    pub fn new_cost_document(
        posting_date: NaiveDate,
        doc_no: String,
        material: Material,
        vendor: Vendor,
        amount: Decimal,
        updated_by: User,
    ) -> Document {
        let lines = vec![
            DocumentLine {
                account: INVENTORY_DIFF_ACCOUNT,
                material: Some(material),
                business_partner: None,
                debt_credit: DebtCredit::Debt,
                amount,
                quantity: None,
            },
            DocumentLine {
                account: PAYABLE_ACCOUNT,
                material: None,
                business_partner: Some(BusinessPartner::Vendor(vendor)),
                debt_credit: DebtCredit::Credit,
                amount: -amount,
                quantity: None,
            },
        ];

        Document {
            posting_date,
            doc_no,
            updated_by,
            lines,
        }
    }

    pub fn new_cogs_document(
        posting_date: NaiveDate,
        doc_no: String,
        customer: Customer,
        material: Material,
        std_price: Decimal,
        quantity: Decimal,
        updated_by: User,
    ) -> Document {
        let std_amount = std_price * quantity;

        let lines = vec![
            DocumentLine {
                account: COGS_ACCOUNT,
                material: Some(material),
                business_partner: Some(BusinessPartner::Customer(customer)),
                debt_credit: DebtCredit::Debt,
                amount: std_amount,
                quantity: Some(quantity),
            },
            DocumentLine {
                account: INVENTORY_ACCOUNT,
                material: Some(material),
                business_partner: None,
                debt_credit: DebtCredit::Credit,
                amount: -std_amount,
                quantity: Some(quantity),
            },
        ];

        Document {
            posting_date,
            doc_no,
            updated_by,
            lines,
        }
    }
    pub fn new_sale_document(
        posting_date: NaiveDate,
        doc_no: String,
        customer: Customer,
        material: Material,
        price: Decimal,
        quantity: Decimal,
        updated_by: User,
    ) -> Document {
        let amount = price * quantity;

        let lines = vec![
            DocumentLine {
                account: RECEIVABLE_ACCOUNT,
                material: None,
                business_partner: Some(BusinessPartner::Customer(customer)),
                debt_credit: DebtCredit::Debt,
                amount,
                quantity: None,
            },
            DocumentLine {
                account: SALES_ACCOUNT,
                material: Some(material),
                business_partner: Some(BusinessPartner::Customer(customer)),
                debt_credit: DebtCredit::Credit,
                amount: -amount,
                quantity: Some(quantity),
            },
        ];

        Document {
            posting_date,
            doc_no,
            updated_by,
            lines,
        }
    }
}

pub struct DocumentLine {
    pub account: &'static str,
    pub business_partner: Option<BusinessPartner>,
    pub material: Option<Material>,
    pub debt_credit: DebtCredit,
    pub amount: Decimal,
    pub quantity: Option<Decimal>,
}

File: golang/model/domain/document.go
package domain

import (
	"cloud.google.com/go/civil"
	"github.com/shopspring/decimal"
)

type Document struct {
	PostingDate civil.Date
	DocNo       string
	UpdatedBy   User
	Lines       []DocumentLine
}

type DocumentLine struct {
	Account         string
	BusinessPartner *string
	Material        *string
	DebtCredit      string
	Amount          decimal.Decimal
	Quantity        *decimal.Decimal
}

func NewPurchaseDocument(
	postingDate civil.Date,
	docNo string,
	material Material,
	vendor Vendor,
	price decimal.Decimal,
	stdPrice decimal.Decimal,
	quantity decimal.Decimal,
	updatedBy User,
) *Document {

	amount := price.Mul(quantity)
	stdAmount := stdPrice.Mul(quantity)
	diffAmount := amount.Sub(stdAmount)
	businessPartner := vendor.Id()
	materialId := material.Id()
	lines := make([]DocumentLine, 0, 3)
	lines = append(lines,
		DocumentLine{
			Account:         INVENTORY_ACCOUNT,
			Material:        &materialId,
			BusinessPartner: nil,
			DebtCredit:      DEBT,
			Amount:          amount,
			Quantity:        &quantity,
		},
	)

	if diffAmount.Equal(decimal.NewFromInt(0)) {
		debtCredit := CREDIT
		if diffAmount.GreaterThan(decimal.NewFromInt(0)) {
			debtCredit = DEBT
		}
		diffLine := DocumentLine{
			Account:         INVENTORY_DIFF_ACCOUNT,
			Material:        &materialId,
			BusinessPartner: nil,
			DebtCredit:      debtCredit,
			Amount:          diffAmount,
			Quantity:        nil,
		}
		lines = append(lines, diffLine)
	}

	creditorLine := DocumentLine{
		Account:         PAYABLE_ACCOUNT,
		Material:        nil,
		BusinessPartner: &businessPartner,
		DebtCredit:      CREDIT,
		Amount:          amount.Neg(),
		Quantity:        nil,
	}
	lines = append(lines, creditorLine)
	return &Document{
		PostingDate: postingDate,
		DocNo:       docNo,
		UpdatedBy:   updatedBy,
		Lines:       lines,
	}

}

func NewCostDocument(
	postingDate civil.Date,
	docNo string,
	material Material,
	vendor Vendor,
	amount decimal.Decimal,
	updatedBy User,
) *Document {
	businessPartner := vendor.Id()
	materialId := material.Id()
	lines := []DocumentLine{
		DocumentLine{
			Account:         INVENTORY_DIFF_ACCOUNT,
			Material:        &materialId,
			BusinessPartner: nil,
			DebtCredit:      DEBT,
			Amount:          amount,
			Quantity:        nil,
		},
		DocumentLine{
			Account:         PAYABLE_ACCOUNT,
			Material:        nil,
			BusinessPartner: &businessPartner,
			DebtCredit:      CREDIT,
			Amount:          amount.Neg(),
			Quantity:        nil,
		},
	}
	return &Document{
		PostingDate: postingDate,
		DocNo:       docNo,
		UpdatedBy:   updatedBy,
		Lines:       lines,
	}
}

func NewCogsDocument(
	postingDate civil.Date,
	docNo string,
	customer Customer,
	material Material,
	stdPrice decimal.Decimal,
	quantity decimal.Decimal,
	updatedBy User,
) *Document {
	businessPartner := customer.Id()
	materialId := material.Id()
	stdAmount := stdPrice.Mul(quantity)
	lines := []DocumentLine{
		DocumentLine{
			Account:         COGS_ACCOUNT,
			Material:        &materialId,
			BusinessPartner: &businessPartner,
			DebtCredit:      DEBT,
			Amount:          stdAmount,
			Quantity:        &quantity,
		},
		DocumentLine{
			Account:         INVENTORY_ACCOUNT,
			Material:        &materialId,
			BusinessPartner: nil,
			DebtCredit:      CREDIT,
			Amount:          stdAmount.Neg(),
			Quantity:        &quantity,
		},
	}
	return &Document{
		PostingDate: postingDate,
		DocNo:       docNo,
		UpdatedBy:   updatedBy,
		Lines:       lines,
	}
}

func NewSaleDocument(
	postingDate civil.Date,
	docNo string,
	customer Customer,
	material Material,
	price decimal.Decimal,
	quantity decimal.Decimal,
	updatedBy User,
) *Document {
	amount := price.Mul(quantity)
	businessPartner := customer.Id()
	materialId := material.Id()
	lines := []DocumentLine{
		DocumentLine{
			Account:    RECEIVABLE_ACCOUNT,
			Material:   nil,
			DebtCredit: DEBT,
			Amount:     amount,
			Quantity:   &quantity,
		},
		DocumentLine{
			Account:         SALES_ACCOUNT,
			Material:        &materialId,
			BusinessPartner: &businessPartner,
			DebtCredit:      CREDIT,
			Amount:          amount.Neg(),
			Quantity:        &quantity,
		},
	}
	return &Document{
		PostingDate: postingDate,
		DocNo:       docNo,
		UpdatedBy:   updatedBy,
		Lines:       lines,
	}

}
